inst/include/dust/kernels.hpp

```diff
@@ -18,18 +18,29 @@ DEVICE void update_device(size_t step,
 
 // __global__ for shuffling particles
 template<typename real_t>
-KERNEL void scatter_device(int* scatter_index,
+KERNEL void scatter_device(const size_t* index,
                            real_t* state,
                            real_t* scatter_state,
-                           size_t state_size) {
+                           const size_t n_state,
+                           const size_t n_particles) {
+  // e.g. 4 particles with 3 states ABC stored on device as
+  // [1_A, 2_A, 3_A, 4_A, 1_B, 2_B, 3_B, 4_B, 1_C, 2_C, 3_C, 4_C]
+  // e.g. index [3, 1, 3, 2] with would be
+  // [3_A, 1_A, 3_A, 2_A, 3_B, 1_B, 3_B, 2_B, 3_C, 1_C, 3_C, 2_C]
+  // interleaved, i.e. input repeated n_state_full times, plus a strided
+  // offset
+  // [3, 1, 3, 2, 3 + 4, 1 + 4, 3 + 4, 2 + 4, 3 + 8, 1 + 8, 3 + 8, 2 + 8]
+  // [3, 1, 3, 2, 7, 5, 7, 6, 11, 9, 11, 10]
+  int state_size = n_state * n_particles;
```

Comment:
out of curiosity what is the motivation for moving to `int` rather than `size_t` here? (with the effect on line 40 with the for loop)

Reply:
A micro-optimisation: https://docs.nvidia.com/cuda/cuda-c-best-practices-guide/index.html#loop-counters-signed-vs-unsigned
Though maybe this could have stayed as size_t (and I guess strictly should be const)

Reply:
> Best practices suggest that this optimization be performed after all higher-level optimizations have been completed 

:joy: